src/query/pipeline/transforms/src/processors/transforms/transform_pipeline_helper.rs

```diff
@@ -0,0 +1,158 @@
+use databend_common_pipeline_core::processors::ProcessorPtr;
+use databend_common_pipeline_core::Pipeline;
+use databend_common_pipeline_core::TransformPipeBuilder;
+
+use crate::processors::AccumulatingTransform;
+use crate::processors::AccumulatingTransformer;
+use crate::processors::AsyncAccumulatingTransform;
+use crate::processors::AsyncAccumulatingTransformer;
+use crate::processors::AsyncTransform;
+use crate::processors::AsyncTransformer;
+use crate::processors::Transform;
+use crate::processors::Transformer;
+
+pub trait TransformPipelineHelper {
+    fn try_add_transform_with_builder<F, R>(
+        &mut self,
+        f: F,
+        create: impl Fn(Arc<InputPort>, Arc<OutputPort>, R) -> Box<dyn Processor>,
+    ) -> Result<()>
+    where
+        F: Fn() -> Result<R>;
+
+    fn try_add_transformer<F, R>(&mut self, f: F) -> databend_common_exception::Result<()>
+    where
+        F: Fn() -> Result<R>,
+        R: Transform + 'static,
+    {
+        self.try_add_transform_with_builder(f, Transformer::<R>::create)
+    }
+
+    fn add_transformer<F, R>(&mut self, f: F)
+    where
+        F: Fn() -> R,
+        R: Transform + 'static,
+    {
+        // Safe to unwrap, since the closure always return Ok(_).
+        self.try_add_transformer(|| Ok(f())).unwrap()
+    }
+
+    fn try_add_async_transformer<F, R>(&mut self, f: F) -> databend_common_exception::Result<()>
+    where
+        F: Fn() -> Result<R>,
+        R: AsyncTransform + 'static,
+    {
+        self.try_add_transform_with_builder(f, AsyncTransformer::<R>::create)
+    }
+
+    fn add_async_transformer<F, R>(&mut self, f: F)
+    where
+        F: Fn() -> R,
+        R: AsyncTransform + 'static,
+    {
+        // Safe to unwrap, since the closure always return Ok(_).
+        self.try_add_async_transformer(|| Ok(f())).unwrap()
+    }
+
+    fn try_add_accumulating_transformer<F, R>(
+        &mut self,
+        f: F,
+    ) -> databend_common_exception::Result<()>
+    where
+        F: Fn() -> Result<R>,
+        R: AccumulatingTransform + 'static,
+    {
+        self.try_add_transform_with_builder(f, AccumulatingTransformer::<R>::create)
+    }
+
+    fn add_accumulating_transformer<F, R>(&mut self, f: F)
+    where
+        F: Fn() -> R,
+        R: AccumulatingTransform + 'static,
+    {
+        self.try_add_accumulating_transformer(|| Ok(f())).unwrap()
+    }
+
+    fn try_add_async_accumulating_transformer<F, R>(
+        &mut self,
+        f: F,
+    ) -> databend_common_exception::Result<()>
+    where
+        F: Fn() -> Result<R>,
+        R: AsyncAccumulatingTransform + 'static,
+    {
+        self.try_add_transform_with_builder(f, AsyncAccumulatingTransformer::<R>::create)
+    }
+
+    fn add_async_accumulating_transformer<F, R>(&mut self, f: F)
+    where
+        F: Fn() -> R,
+        R: AsyncAccumulatingTransform + 'static,
+    {
+        self.try_add_async_accumulating_transformer(|| Ok(f()))
+            .unwrap()
```

Comment:
```suggestion
            ?
```

Reply:
the method returns `()`, question mark does not work here.
 ~~~
 fn add_async_accumulating_transformer<F, R>(&mut self, f: F)
    where
        F: Fn() -> R,
        R: AsyncAccumulatingTransform + 'static,
        
~~~        

but, I should left a comment for it , like `fn add_async_transformer` does